confuse.py

```diff
@@ -134,36 +135,193 @@ def __init__(self, filename, reason=None):
+        If it doesn't load, return False to keep it marked as unloaded.
+        Otherwise it will be assumed to be loaded.
+        '''
+
+    def config_dir(self, create=True):
+        '''Create the config dir, if there's a filename associated with the
+        source.'''
+        if self.filename:
+            dirname = os.path.dirname(self.filename)
+            if create and dirname and not os.path.isdir(dirname):
+                os.makedirs(dirname)
+            return dirname
+        return None
 
     @classmethod
-    def of(cls, value):
-        """Given either a dictionary or a `ConfigSource` object, return
-        a `ConfigSource` object. This lets a function accept either type
-        of object as an argument.
-        """
+    def is_of_type(cls, value):
+        '''Determine if value is a valid parameter for this Source.'''
+        return
+
+    # overriding dict methods so that the configuration is loaded before any
+    # of them are run
+    __getitem__ = _load_first(dict.__getitem__)
+    __iter__ = _load_first(dict.__iter__)
+    # __len__ = _load_first(dict.__len__)
+    keys = _load_first(dict.keys)
+    values = _load_first(dict.values)
+
+    @classmethod
+    def of(cls, value, **kw):
+        '''Try to convert value to a `ConfigSource` object. This lets a
+        function accept values that are convertable to a source.
+        '''
+        # ignore if already a source
         if isinstance(value, ConfigSource):
             return value
-        elif isinstance(value, dict):
-            return ConfigSource(value)
-        else:
-            raise TypeError(u'source value must be a dict')
+
+        # convert using one of the configsource subtypes
+        for subcls in all_subclasses(ConfigSource):
+            if subcls.is_of_type(value):
+                return subcls(value, **kw)
+
+        # if none matched, use convert dict to ConfigSource
+        if isinstance(value, dict):
+            return ConfigSource(value, **kw)
+        raise TypeError(
+            u'ConfigSource.of value unable to cast to ConfigSource.')
+
+
+class YamlSource(ConfigSource):
+    '''A config source pulled from yaml files.'''
+    EXTENSIONS = '.yaml', '.yml'
+
+    def __init__(self, filename=None, value=UNSET, default=False,
+                 ignore_missing=False):
+        self.ignore_missing = ignore_missing
+        super(YamlSource, self).__init__(value, filename, default)
+
+    @classmethod
+    def is_of_type(cls, value):
+        '''Does value look like a .yaml file?'''
+        return (
+            isinstance(value, BASESTRING)
+            and os.path.splitext(value)[1] in cls.EXTENSIONS)
+
+    _loader = lambda self, f: load_yaml(f)
+
+    def _load(self):
+        '''Load the file if it exists.'''
+        if self.ignore_missing and not os.path.isfile(self.filename):
+            return False
+        self.update(self._loader(self.filename))
+
+
+class EnvSource(ConfigSource):
```

Comment:
This seems pretty cool! But it's a separate feature—maybe we can discuss it separately? (We can also look into writing some documentation for the new variable-based route to configuration…)

Reply:
yeah we can move this out. It just seemed like a simple feature made simpler by this PR, so I added it as a demonstration of source extensibility. but it can be added later.